CONTRIBUTING.md

```diff
@@ -800,45 +744,280 @@ Try restricting ifdefs in the header files, or in the beginning of
+This is what you use for the pull request there.
+Pushing more commits or force pushing changed commits to the same branch will
+automatically update an open Pull Request which uses that branch.
+
+
+### Cleaning up the commit history - rebasing and squashing
+
 
-#### How to rebase and squash commits
+#### Resolving conflicts using rebase
 
-[This post has a good explanation.](https://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit)
+After you branched from `upstream/master` it is possible that someone else
+(or even you) commit conflicting changes to the `master` branch.
+To resolve these conflicts in your PR branch, use:
+
+```
+$ git fetch --all
+$ git rebase upstream/master
+```
+
+This will stop at the conflicting commit, and give you instructions to
+proceed. You will have to resolve conflicts and then run:
+
+```
+$ git rebase --continue
+```
+
+This might happen multiple times (1 for each commit).
+Once done, you will have to force push your changes to the PR branch:
+
+```
+$ git push -f
+```
+
+
+#### Interactive rebase
+
+We want the commit history to be as easy to read and understand as possible.
+Sometimes we make more commits than necessary, and we want to squash them
+(combine them). A good tool for this is _interactive rebase_.
+
+On your PR branch, run:
+
+```
+$ git rebase -i upstream/master
+```
+
+This will open an editor (usually `vi` or `emacs`), with a list of your
+commits:
+
+```
+pick ac07e6d27 Added new policy function isrealdir()
+pick 001bd420b Fixed mistake in isrealdir()
+
+# Commented out instructions ....
+```
+
+`upstream/master` will be used as a starting point, and the commits / commands
+will be applied tofrom top to bottom of this list. To combine the 2 commits,
+replace `pick` with `squash` on the second line:
+
+```
+pick ac07e6d27 Added new policy function isrealdir()
+squash 001bd420b Fixed mistake in isrealdir()
+
+# Commented out instructions ....
+```
+
+This will squash the second commit into the first one (the one above).
+Save and close. Another editor will open, allowing you to update the commit
+message(s). Update it to reflect the current state of this commit, then save
+and close.
+
+Once you get comfortable with _interactive rebase_ you can make more advanced
+changes, like squashing multiple commits, reordering commits, or dropping
+some commits. You may create conflicts, that you have to resolve, as
+described above.
+
+See also:
+[Squashing GitHub Pull Requests into a single commit](https://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit)
+
+### Keeping master up to date
+
+Switch to local master branch:
+
+```
+$ git checkout master
+```
+
+Download changes from all remotes:
+
+```
+$ git fetch --all
+```
+
+Apply changes from the `upstream` `master` branch to our local `master` branch:
+
+```
+$ git rebase upstream/master
+```
+
+(You can also use `git pull`, but I find `rebase` a little easier when there
```

Comment:
maybe `git pull --rebase`

Reply:
Sure. But I still think it's easier to just learn `fetch` + `rebase`, and after @jimis taught me long ago, I almost never use `pull` or `merge` commands. (This is of course very subjective, there's no right answer, I just tried to show my preferred approach, both now and when I learned git).